typescript-loader/loader.js

```diff
@@ -0,0 +1,73 @@
+// 💡 It's probably a good idea to put this loader near the end of the chain.
+
+import path from 'node:path';
+
+import { transform } from 'esbuild';
+
+const jsToTs = new Map([
+  // Sorted by priority (first wins)
+  ['.js',  '.ts' ],
+  ['.cjs', '.cts'],
+  ['.mjs', '.mts'],
+  ['.jsx', '.tsx'],
+]);
+const jsExts = new Set(jsToTs.keys());
+const tsExts = new Set(jsToTs.values());
+
+export function resolve(specifier, context, nextResolve) {
+  const { base, ext, name } = path.parse(specifier);
+
+  if (tsExts.has(ext)) { // No guessing needed
+    const { url } = nextResolve(specifier, context);
+
+    return {
+      format: 'typescript', // Provide a signal to `load`
+      shortCircuit: true,
+      url,
+    };
+  }
+
+  if (!jsExts.has(ext)) { // When file is not ts or js, it's irrelevant
+    return nextResolve(specifier, context);
+  }
+
+  try { // Check whether such a js file does exist
+    return nextResolve(specifier, context);
+  } catch (err) {
+    if (err?.code !== "ERR_MODULE_NOT_FOUND") { throw err; }
+  }
+
+  for (const tsExt of tsExts) { // Finally, check whether any ts file exists
```

Comment:
IIUC, that's not how TypeScript does it, nor anyone else I think. Importing a `.mjs` specifier can only match to a `.mts`, `.cjs` can only match `.cts`, and `.ts` `.js`. I don't think it's correct to try all extensions here, there should be only one.